### PR TITLE
Feature/lightning sequence labeling pipeline

### PR DESCRIPTION
**Pull request by djaniak**

### CODE REVIEW

File: embeddings/data/datamodule.py [hunk 1]
```diff
@@ -195,3 +222,108 @@ def convert_to_features(
...
+            )
+        self.num_classes = len(labels)
+        for i, label in enumerate(labels):
+            self.label_to_id[label] = i
+            self.id_to_label[i] = label
+
+    def convert_to_features(
+        self, example_batch: Dict[str, Any], indices: Optional[List[int]] = None
+    ) -> BatchEncoding:
+        texts = example_batch[self.text_field]
```
Comment (asawczyn) [inline embeddings/data/datamodule.py:286]: `texts` -> `text` because it is uncountable

File: embeddings/data/data_collator.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+from dataclasses import dataclass
+from typing import Any, Dict, List, Union
+
+import torch
+from transformers import BatchEncoding, DataCollatorForTokenClassification
+
+
+# ignoring `Class cannot subclass "DataCollatorForTokenClassification" (has type "Any")` error
+@dataclass
+class CustomDataCollatorForTokenClassification(DataCollatorForTokenClassification):  # type: ignore
```
Comment (asawczyn) [inline embeddings/data/data_collator.py:10]: where is a difference in the logic of the custom one? I have read the original version and i do not see it https://github.com/huggingface/transformers/blob/19e5ed736611227b004c6f55679ce3536db3c28d/src/transformers/data/data_collator.py#L263

**Review by asawczyn — CHANGES_REQUESTED** [review]

**Comment by asawczyn** [comment]
just few comments, I have not finish my reading yet

File: embeddings/data/data_collator.py (hunk 2 above)
Comment (djaniak) [reply embeddings/data/data_collator.py:10]: i removed the labels from dictionary that is passed to tokenizer for padding because it does not pad the labels and when the tokenizer returns BatchEncoding object it tries to convert the whole dict to tensor and it fails

File: embeddings/task/lightning_task/sequence_labeling.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        assert self.trainer is not None
+        for i, (pred, gt) in enumerate(zip(predictions, ground_truth)):
+            predictions[i] = [
+                self.trainer.datamodule.id_to_label[x] for x in list(pred[gt != self.ignore_index])
+            ]
+            ground_truth[i] = [
+                self.trainer.datamodule.id_to_label[x] for x in list(gt[gt != self.ignore_index])
+            ]
+
+        return {"y_pred": np.array(predictions), "y_true": np.array(ground_truth)}
```
Comment (asawczyn) [inline embeddings/task/lightning_task/sequence_labeling.py:129]: Are conversions in lines 117 and 118 necessary? you can iterate over torch objects

**Review by asawczyn — CHANGES_REQUESTED** [review]

File: embeddings/evaluator/sequence_labeling_evaluator.py [hunk 4]
```diff
@@ -64,3 +64,7 @@ def metrics(
         self,
     ) -> Sequence[Metric[Union[List[Any], nptyping.NDArray[Any], torch.Tensor], Dict[Any, Any]]]:
         return [self.metric]
+
+
+EvaluationMode = SequenceLabelingEvaluator.EvaluationMode
+TaggingScheme = SequenceLabelingEvaluator.TaggingScheme
```
[the remark this answers is not in the document]
Comment (laugustyniak) [reply embeddings/evaluator/sequence_labeling_evaluator.py:70]: +1

**Review by laugustyniak — APPROVED** [review]

File: embeddings/data/datamodule.py [hunk 5]
```diff
@@ -195,3 +222,108 @@ def convert_to_features(
...
+        eval_batch_size: int = 32,
+        label_all_tokens: bool = False,
+        tokenizer_kwargs: Optional[Dict[str, Any]] = None,
+        batch_encoding_kwargs: Optional[Dict[str, Any]] = None,
+        load_dataset_kwargs: Optional[Dict[str, Any]] = None,
+        **kwargs: Any,
+    ):
+        self.text_field = text_field
+        self.label_all_tokens = label_all_tokens
+        self.batch_encoding_kwargs = (
```
[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/data/datamodule.py:252]: it will use the default parameters from the Tokenizer object (precisely from __call__ method)

File: embeddings/evaluator/sequence_labeling_evaluator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/evaluator/sequence_labeling_evaluator.py:70]: cannot access an object's attribute before declaration unless there is other way to do it

File: embeddings/data/datamodule.py [hunk 6]
```diff
@@ -195,3 +222,108 @@ def convert_to_features(
...
+        else:
+            # Create unique label set from train dataset.
+            labels = sorted(
+                {label for column in self.dataset["train"][self.target_field] for label in column}
+            )
+        self.num_classes = len(labels)
+        for i, label in enumerate(labels):
+            self.label_to_id[label] = i
+            self.id_to_label[i] = label
+
```
[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/data/datamodule.py:282]: do you have any examples of datasets that for sequence labeling that do not have a label column in  datasets.Sequence(datasets.features.ClassLabel(...) type so I can work this out? kpwr-ner and nkjp-pos from clarin are both already converted to this type

[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/data/datamodule.py:282]: i've looked into our example for flair sequence labeling and we do not handle the case where the label column is not Squence(ClassLabel) type

casting could be done like this:

```
labels = sorted(
                {label for column in self.dataset["train"][self.target_field] for label in column}
            )
            new_features = self.dataset["train"].features.copy()
            new_features['labels'] = Sequence(
                feature=ClassLabel(num_classes=len(labels), names=list(labels)))
            self.dataset = self.dataset.cast(new_features)
```

but I was thinking whether we should handle this in our code for lightning dataset

File: embeddings/task/lightning_task/sequence_labeling.py (hunk 3 above)
Comment (djaniak) [reply embeddings/task/lightning_task/sequence_labeling.py:129]: yes, I can iterate over numpy object but I cannot do the replacement of values in numpy array (predictions[I] = ...)
nevertheless, I refactored the code little bit here

File: embeddings/data/datamodule.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/data/datamodule.py]: this is sorted out now (dataset is casted again after the labels are converted) and error is raised when the dataset is not Sequence(ClassLabel(...)) as in our flair corpus

[the remark this answers is not in the document]
Comment (asawczyn) [reply embeddings/data/datamodule.py]: meh, I agree that users should be responsible for preparing the dataset. The prepared code for casting does not work, it is caused by the fact, that `.cast() ` written by HF does not convert the column. It tries to read column as if it were written in `ClassLabel` i.e. integers.


```from datasets import Dataset, ClassLabel
from datasets import Sequence as DatasetsSequence

my_dict = {"t": [["a", "a", "o"], ["b", "o", "a"], ]}
dataset = Dataset.from_dict(my_dict)


new_features = dataset.features.copy()
new_features["t"] = DatasetsSequence(
        feature=ClassLabel(num_classes=3, names=["a", "b", "o"])
    )
dataset = dataset.cast(new_features)```
output: `pyarrow.lib.ArrowInvalid: Failed to parse string: 'a' as a scalar of type int64`

Because of that, though I preferred the first version or the second with raising the error, the casting can be removed.

[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/data/datamodule.py]: yes, I am aware of the fact that it only works if the labels are integers, but it is the default way to encode the labels in HF from what I see; now we only cast the labels because they are put in Values when calling the convert_to_features method - we just encode them again as labels as they were before. This way we can use the int2str and str2int function for sequence labeling

**Comment by ktagowski** [comment]
Mypy issue separated #152

**Review by asawczyn — CHANGES_REQUESTED** [review]

File: embeddings/data/datamodule.py [hunk 7]
```diff
@@ -107,47 +103,77 @@ def prepare_data(self) -> None:
     def setup(self, stage: Optional[str] = None) -> None:
         self.dataset = self.load_dataset()
         self.downsample_dataset()
-        self.num_classes = self.get_num_classes()
+        self.prepare_labels()
         self.process_data()
```
[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/data/datamodule.py:107]: Yes. First we fix the column before called to fix the column before getting the labels information and afterwards to fix it to ClassLabel again after calling convert_to_features which leave it as Value(...) type

File: embeddings/data/datamodule.py [hunk 8]
```diff
@@ -107,47 +103,77 @@ def prepare_data(self) -> None:
...
+        "truncation": True,
+    }
+
     def __init__(
         self,
         tokenizer_name_or_path: str,
         dataset_name: str,
         text_fields: Union[str, Sequence[str]],
         target_field: str,
-        max_seq_length: Optional[int] = None,
```
[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/data/datamodule.py:150]: I did not want to double the parameters with HuggingFaceDataModule and you could add this (and other) argument to HuggingFaceDataModule through keyword arguments. Do you think it will be confusing to the user? If so I can change it back the way it was previosuly.

[the remark this answers is not in the document]
Comment (asawczyn) [reply embeddings/data/datamodule.py:150]: I am not a fan of passing args through kwargs, but I see there are many of them, so ok.

File: embeddings/task/lightning_task/text_classification.py [hunk 9]
```diff
@@ -88,27 +87,25 @@ def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
...
         return None
 
     def predict_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
         batch, batch_idx = args
-        loss, preds = self.shared_step(**batch)
-        if isinstance(preds, SequenceClassifierOutput):
-            assert isinstance(preds.logits, torch.Tensor)
-            return preds.logits
+        loss, logits = self.shared_step(**batch)
+        preds = torch.argmax(logits, dim=1)
```
Comment (ktagowski) [inline embeddings/task/lightning_task/text_classification.py:108]: I'm not a fan of that change. As in some cases sometimes scores may be needed

File: embeddings/data/datamodule.py [hunk 10]
```diff
@@ -124,38 +136,42 @@ def downsample_dataset(self) -> None:
...
             self.convert_to_features,
             batched=True,
+            batch_size=self.processing_batch_size,
             remove_columns=columns,
         )
+        self._class_encode_column("labels")
         self.dataset.set_format(type="torch")
 
     def train_dataloader(self) -> DataLoader[HuggingFaceDataset]:
-        return DataLoader(self.dataset["train"], batch_size=self.train_batch_size, shuffle=True)  # type: ignore
```
Comment (ktagowski) [inline embeddings/data/datamodule.py:147]: Why shuffle argument removed here?

File: embeddings/data/datamodule.py [hunk 11]
```diff
@@ -124,38 +136,42 @@ def downsample_dataset(self) -> None:
...
+            dataset=self.dataset["train"],  # type: ignore
+            batch_size=self.train_batch_size,
+            collate_fn=self.collate_fn,
+        )
 
     # Ignoring the type of val_dataloader method from supertype "DataHooks" allowing for None
     # and training without validation dataset.
     def val_dataloader(self) -> Optional[DataLoader[HuggingFaceDataset]]:  # type: ignore
         if "validation" in self.dataset:
-            return DataLoader(self.dataset["validation"], batch_size=self.eval_batch_size, shuffle=False)  # type: ignore
```
Comment (ktagowski) [inline embeddings/data/datamodule.py:153]: Why shuffle argument removed here?

File: embeddings/data/datamodule.py [hunk 12]
```diff
@@ -124,38 +136,42 @@ def downsample_dataset(self) -> None:
...
+            return DataLoader(
+                dataset=self.dataset["validation"],  # type: ignore
+                batch_size=self.eval_batch_size,
+                collate_fn=self.collate_fn,
+            )
         else:
             return None
 
     def test_dataloader(self) -> DataLoader[HuggingFaceDataset]:
-        return DataLoader(self.dataset["test"], batch_size=self.eval_batch_size, shuffle=False)  # type: ignore
```
Comment (ktagowski) [inline embeddings/data/datamodule.py:158]: Why shuffle argument removed here?

**Review by ktagowski — CHANGES_REQUESTED** [review]

File: embeddings/data/datamodule.py (hunk 10 above)
Comment (djaniak) [reply embeddings/data/datamodule.py:147]: happened accidentally during the rebase, I'll fix it

File: embeddings/data/datamodule.py [hunk 13]
```diff
@@ -124,38 +136,45 @@ def downsample_dataset(self) -> None:
...
     # Ignoring the type of val_dataloader method from supertype "DataHooks" allowing for None
     # and training without validation dataset.
     def val_dataloader(self) -> Optional[DataLoader[HuggingFaceDataset]]:  # type: ignore
         if "validation" in self.dataset:
-            return DataLoader(self.dataset["validation"], batch_size=self.eval_batch_size, shuffle=False)  # type: ignore
+            return DataLoader(
+                dataset=self.dataset["validation"],  # type: ignore
+                batch_size=self.eval_batch_size,
+                collate_fn=self.collate_fn,
+                shuffle=True,
```
Comment (ktagowski) [inline embeddings/data/datamodule.py:166]: `shuffle=False` for validation dataloader

Comment (ktagowski) [inline embeddings/data/datamodule.py:166]:
```suggestion
                shuffle=False,
```

File: embeddings/data/datamodule.py [hunk 14]
```diff
@@ -124,38 +136,45 @@ def downsample_dataset(self) -> None:
...
         else:
             return None
 
     def test_dataloader(self) -> DataLoader[HuggingFaceDataset]:
-        return DataLoader(self.dataset["test"], batch_size=self.eval_batch_size, shuffle=False)  # type: ignore
+        return DataLoader(
+            dataset=self.dataset["test"],  # type: ignore
+            batch_size=self.eval_batch_size,
+            collate_fn=self.collate_fn,
+            shuffle=True,
```
Comment (ktagowski) [inline embeddings/data/datamodule.py:176]: Test dataloader shouldn't be shuffled
```suggestion
            shuffle=False,
```

**Review by ktagowski — APPROVED** [review]

File: embeddings/data/datamodule.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (laugustyniak) [reply embeddings/data/datamodule.py:298]: I meant that part will be passed explicitly, and part could be different than defaults here ```DEFAULT_BATCH_ENCODING_KWARGS = {
        "padding": True,
        "truncation": True,
        "is_split_into_words": True,
        "return_offsets_mapping": True,
    }```

**Review by laugustyniak — APPROVED** [review]

File: embeddings/data/datamodule.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/data/datamodule.py:298]: if somebody passes batch_encoding_kwargs they will automatically overwrite the DEFAULT_BATCH_ENCODING_KWARGS so he has to add the remaining keys and values to the dictionary with these kwargs. It is done to allow the maximum level of freedom to the user however in most cases the DEFAULT_BATCH_ENCODING_KWARGS is sufficient.

**Review by asawczyn — APPROVED** [review]
